### PR TITLE
Legends

### PR DESCRIPTION
**Pull request by Casyfill**
Hi there. Here is the alteration that seems to pass your tests and fixes the legends.
In particular

- circle style for `_paint_carto_legend` is now consistent with the plot itself (edge colors, face colors)
- by default Sankey lines have `round` end filling
- Sankey now renders a new type of legend, `_paint_sankey_legend`  that draws lines, also style-consistent.
- Legends now add legend artist to the plot on the contrary to calling ax.Legend.

A few questions/warnings
- I don't mean this PR to be accepted (at this time), but rather to be a place for you to express your concerns/opinions

1. I didn't get why you plot markers with `20*(dscale(x))**(1/2)` in `_paint_carto_legend`.  As I "fixed" markersize in the N1, it might now affect your example images.
2. while adding artists fixed an issue with double legend for me, it might affect plotting multiple subplots. To my shape, I didn't check that yet. Still, I think there should be a way to have multiple legends on multiple subplots.
3. As I plot network, I'd like to have one scaling factor both for nodes (points) and edges (sankey). Currently the only way to do that is to calculate upper bound from one using data from another.

4. Generally, do you think if it makes sense to add yet another type of plot - [geo]network? and handle everything with nodes and legends there?

### CODE REVIEW

**Comment by Casyfill** [comment]
overal_map2

**Comment by ResidentMario** [comment]
It looks to me like your text editor is set to force break lines past 80 character, which I don't do. It's not really necessary and it's causing a lot of messing diffing, can you fix that up please?

**Comment by Casyfill** [comment]
Hey there. 
Sure, here I rolled back all the text editor changes - and everything but the part that keeps legend consistent with the map. Let's review all other bullet points after that

File: geoplot/geoplot.py [hunk 1]
```diff
@@ -2532,6 +2532,8 @@ def _paint_carto_legend(ax, values, legend_values, legend_labels, scale_func, le
         Keyword arguments which will be passed to the matplotlib legend instance on initialization. This parameter
         is provided to allow fine-tuning of legend placement at the top level of a plot method, as legends are very
         finicky.
+     map_kwargs : dict
+        Vis properties from the original map.
```
Comment (ResidentMario) [inline geoplot/geoplot.py]: This needs a more descriptive docstring.

**Comment by ResidentMario** [comment]
The changeset in the current diff looks fine, though I will still need to look at it more in-depth. You should split up the rest of your proposed changes into two other PRs: one for the Sankey stuff, and one for the legend stuff.

**Comment by ResidentMario** [comment]
For your questions:

1. I don't remember. Surely it must be because it gives me a reasonable default in a broad range of cases, however.
2. See `4`.
3. See `4`.
4. Not a bad idea. You can get this by overplotting a `pointplot` on a `sankey` (as I assume you did here), and handling the double-scaling internally within the method. I think that this plot would be the ideal place to try to handle double legends as well. Once that's working, it could then get carried over to [cartograms](http://www.residentmar.io/geoplot/cartogram.html#geoplot.cartogram) as well.

**Comment by Casyfill** [comment]
sounds good, will add more info to the docstring soon

File: src/pip-delete-this-directory.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+This file is placed here by pip to indicate the source was put
```
Comment (ResidentMario) [inline src/pip-delete-this-directory.txt]: `gitignore` this. :)

**Comment by Casyfill** [comment]
done

**Comment by ResidentMario** [comment]
Sorry, thought I'd already merged this. Will get to this later today. I will goal getting the rest tested in a week of proposal, once you get those PRs going.

**Comment by ResidentMario** [comment]
I'm testing this on one of the plots in the docs, and I'm not seeing the intended behavior.

Code:

```python
import sys; sys.path.insert(0, '../')  # ignore this
import geoplot as gplt

%matplotlib inline
import geopandas as gpd
import geoplot.crs as gcrs

cities = gpd.read_file("../data/cities/citiesx010g.shp")
continental_cities = cities[cities['STATE'].map(lambda s: s not in ['PR', 'AK', 'HI', 'VI'])]
continental_cities = continental_cities[continental_cities['POP_2010'] >= 100000]
usa = gpd.read_file("../data/united_states/usa.geojson")
continental_usa = usa[~usa['adm1_code'].isin(['USA-3517', 'USA-3563'])]

proj = gcrs.AlbersEqualArea(central_longitude=-98, central_latitude=39.5)

ax = gplt.polyplot(continental_usa, 
                   projection=proj,
                   zorder=-1,
                   linewidth=0.5,
                   figsize=(12, 12))
gplt.pointplot(continental_cities, 
               scale='POP_2010',
               limits=(1, 80),
               linewidth=0.5,
               edgecolor='black',
               legend=True,  # Add a legend.
               legend_values=[8000000, 6000000, 4000000, 2000000, 100000],
               legend_labels=['8 million', '6 million', '4 million', '2 million', '100 thousand'],
               # Change the legend display to better fit our values.
               legend_kwargs={'frameon': False, 'loc': 'lower right'},
               # Turn off the legend frame and pin it to the bottom right.
               ax=ax, projection=proj)

ax.set_ylim((-1597757, 1457718)); pass
```

Nets:

image

Whereas of course we want to have crisp black-border/steelblue-fill circles, as in the main chart body.

The other thing is that I think you should include `markerfacecolor` and whatever the border argument is as possible keyword arguments for `legend_kwargs` for these plots. The reason being that in this case, and I think in many others, the circles in the legend are going to occlude one another, so we would want to give the user the power to e.g. give the circles a transparent fill anyway (as would be most appropriate here).

**Comment by Casyfill** [comment]
Hm, that worked for me.

Will look into that, thanks!
Regarding the transparency, you might be right, let me think through it.

**Comment by ResidentMario** [comment]
Closing as stale.

**Pull request closed without merging**